Introduce tool manager (#24)

planger opened the pull request:
A tool manager keeps a list of available tools and manages their state. The tool manager itself is controlled via an action handler that reacts to dedicated actions (e.g. EnableToolsAction), so the enablement of tools can easily be controlled by sending those actions. On each action enablement or disablement, the tool manager initializes or disposes the respective tools, which allows them to register and unregister their listeners, as well as to perform any arbitrary diagram modifications, etc., such as setting it up for visual feedback in the diagram, etc. that can later be cleaned up when the tool is disabled. To accommodate for the typical scenarios, the tool manager supports two types of tools, standard tools and "other" tools. Standard tools become active in the common user-interaction state until other tools are explicitly activated. Once a tool is finished or the user escapes the current interaction mode, all standard tools can be activated again.

See #24

spoenemann commented on src/base/tool-manager/tool-manager.ts:
```diff
@@ -0,0 +1,143 @@
+ *
+ * Contributors:
+ * 	Philip Langer - initial API and implementation
+ ******************************************************************************/
```
Please use the following copyright text for all new files:
```
/********************************************************************************
 * Copyright (c) 2018 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
```

planger commented on src/base/tool-manager/tool-manager.ts:
```diff
@@ -0,0 +1,143 @@
+    readonly managedTools: Tool[];
+
+    /** The tools that are enabled by standard, whenever no other tool is enabled. */
+    readonly standardTools: Tool[];
```
Yes, I agree. I think I originally avoided the term default to distinguish it from the `DefaultToolManager` implementation, but since we should put an `I` prefix before the interface anyway, we avoid the name clash and can call the default implementation simply `ToolManager`.

planger commented on src/base/tool-manager/tool-manager.ts:
```diff
@@ -0,0 +1,143 @@
+    enableStandardTools(): void;
+
+    /** Disables all currently active tools. After this call, no tool will be active anymore. */
+    disableActiveTools(): void;
```
My intention was to make this very explicit ( you have to invoke `disableAllTools()` & `enableDefaultTools()`) and allow clients to disable all tools for specific purposes.

But we can change that if you prefer. Then I would limit the API to `enable(toolIds: string[])` for enabling non-default tools and `enableDefaultTools()`, because then there is no need for a `disableActiveTools()` anymore.

JanKoehnlein replied:
Maybe there are two layers of API: The "action" one with the behavior you just described, and the low level one, that lets me selectively activate and deactivate tools. So it would be `enable(toolId: string)` and `disable(toolId: string)` without any side-effects on the other tools. For example, we could additionally activate an edge-edit tool when an edge is selected.

planger commented on src/base/tool-manager/tool-manager.ts:
```diff
@@ -0,0 +1,143 @@
+
+    readonly standardTools: Tool[] = [];
+    readonly tools: Tool[] = [];
+    private actives: Tool[] = [];
```
Ok. I didn't use readonly because of line 81 where we do `this.actives = []`. The available methods on arrays are a bit limited (e.g. there is no `clear()`). But in a subsequent commit, I made actives readonly and splice the array to clear it.

planger commented:
Thanks a lot for the review! I pushed a commit that addresses your comments.

> My only concern is that I don't see yet how multiple active tools would interact with each other (there are multiple standard tools), e.g. a node move tool and a scroll tool which should act on the same events, but exclusively. I like that I currently don't have to switch between tools, but this introduces a strange coupling of these functionalities in the code. Maybe we should be able to prioritize/order tools and provide a better channel than event.preventDefault to communicate that an action has already been taken by a prior tool in the event listener chain.
> 
> This may be out of scope for this issue, but I'd really like to at least have an idea of a comprehensive solution which may involve tools, tool managers, a palette, and some shipped tools like a marque selection tool.

Yes, the tool manager doesn't really solve the problem of interfering tools that are active at the same time. At most, the tool manager mitigates the problem a bit, because we can design the tools in a way that less tools are active at a time (e.g. only active a create-edge tool when the user triggers the creation of the edge).

But in general I agree that a comprehensive mechanism to control potentially interfering tools would be good. I'm not sure what the best way would be though. The first thing that comes to my mind (and what also other frameworks do), is that listeners have a configurable order and that they can optionally "consume" events so that they won't get propagated after a tool consumed them.

If you agree, I could -- based on the tool manager -- migrate the select tool and move tool into tool-manager-compatible tools and implement a marquee tool which would be activated with a keyboard shortcut. What do you think? This way, we have an example of actually using the tool manager in the code base.

JanKoehnlein commented:
Before migrating the existing listeners to tools, let's define the difference between a tool and a listener and find a solution to explicitly manage the order of event processing/consumption.

spoenemann commented:
I'm curious how a Tool implementation would look like. @planger could you point to some existing implementation from the glsp codebase?

planger commented:
> I'm curious how a Tool implementation would look like. @planger could you point to some existing implementation from the glsp codebase?

@spoenemann Sure, here is e.g. the create node and edge tool implementation:
https://github.com/eclipsesource/graphical-lsp/blob/master/client/packages/glsp-sprotty/src/features/tools/creation-tool.ts

Here is a short video of it in action (actually do demonstrate the work in progress palette, but it also shows the node and edge creation, see also https://github.com/eclipsesource/graphical-lsp/pull/154). Note that the feedback is invoked via actions and handled in a [separate module](https://github.com/eclipsesource/graphical-lsp/blob/master/client/packages/glsp-sprotty/src/features/tool-feedback/creation-tool-feedback.ts).
![image](https://user-images.githubusercontent.com/2311075/51808411-bb55a580-2293-11e9-8360-2da84b634022.gif)
